### PR TITLE
fix(compiler): prevent race conditions

### LINKED ISSUE #3206: Regression: "Unconditional component cycle" using DynamicComponentLoader?
Not much to share externally other than stracktrace:

```
Unconditional component cycle in MaterialMenuComponent
STACKTRACE:
package:angular2/src/core/compiler/compiler.dart 227:9  Compiler._compileNestedProtoViews.<fn>
package:stack_trace                                     StackZoneSpecification.registerUnaryCallback.<fn>
package:angular2/src/core/zone/ng_zone.dart 191:40      NgZone._runUnary.<fn>
dart:async                                              _ZoneDelegate.run
package:angular2/src/core/zone/ng_zone.dart 167:21      NgZone._run
package:angular2/src/core/zone/ng_zone.dart 191:7       NgZone._runUnary
dart:async                                              _Future._asyncComplete.<fn>
package:angular2/src/core/zone/ng_zone.dart 200:11      NgZone._scheduleMicrotask.<fn>
dart:async                                              _ZoneDelegate.run
package:angular2/src/core/zone/ng_zone.dart 167:21      NgZone._run
```

Tests didn't catch it, even though we have them. Maybe test environment is subtly different?

### PR DESCRIPTION
**Pull request by tbosch**
Previously, the compiler would detect cycles where there were none just because of other components that were compiled in parallel. Furthermore, the way ProtoView merging was triggered could result into early exits resulting in errors when trying to instantiate ProtoViews.

Closes #3206

### CODE REVIEW

File: modules/angular2/src/core/compiler/compiler.ts [hunk 1]
```diff
@@ -233,67 +206,85 @@ export class Compiler {
     return MapWrapper.values(directivesMap);
   }
 
-  private _compileNestedProtoViews(renderProtoView: renderApi.ProtoViewDto,
-                                   appProtoView: AppProtoView,
-                                   componentType: Type): Promise<AppProtoView> {
+  private _compileNestedProtoViews(appProtoViews: AppProtoView[], componentType: Type,
+                                   componentStack: Map<Type, AppProtoView>): Promise<AppProtoView> {
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/compiler.ts]: componentStack is misleading name, componentPath is more descriptive to what it is collecting.

File: modules/angular2/src/core/compiler/compiler.ts [hunk 2]
```diff
@@ -233,67 +206,85 @@ export class Compiler {
...
+          .forEach((elementBinder: ElementBinder) => {
+            var nestedComponent = elementBinder.componentDirective;
+            var nestedComponentType = <Type>nestedComponent.key.token;
+            var elementBinderDone =
+                (nestedPv: AppProtoView) => { elementBinder.nestedProtoView = nestedPv; };
+            if (componentStack.has(nestedComponentType)) {
+              // cycle...
+              if (appProtoView.isEmbeddedFragment) {
+                throw new BaseException(
+                    `<ng-content> is used within the recursive path of ${stringify(nestedComponentType)}`);
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/compiler.ts]: Ideally we should have a test for this, doesn't seem like a race condition.

File: modules/angular2/src/core/compiler/compiler.ts [hunk 3]
```diff
@@ -233,67 +206,85 @@ export class Compiler {
...
+                elementBinderDone(<AppProtoView>nestedCall);
+              }
+            }
+          });
     });
     return PromiseWrapper.all(nestedPVPromises)
-        .then((_) => {
-          this._collectMergableProtoViews(appProtoView, componentType);
-          return appProtoView;
+        .then(_ => {
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/compiler.ts]: `=> { return foo}` could be simply `=> foo`

**Comment by rkirov** [comment]
In general, compiler code is still quite hard for me to follow. We can improve the documentation and naming of methods and variables once it stabalizes.

File: modules/angular2/src/core/compiler/compiler.ts (hunk 2 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/compiler.ts]: yes

File: modules/angular2/src/core/compiler/compiler.ts (hunk 3 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/compiler.ts]: done

File: modules/angular2/src/core/compiler/compiler.ts (hunk 1 above)
Comment (tbosch) [inline modules/angular2/src/core/compiler/compiler.ts]: done

**Pull request closed without merging**